draft-ietf-gnap-core-protocol.md

```diff
@@ -5213,6 +5213,34 @@ method against the deployment capabilities of the client software and its
 environment. Due to the increased security, an interaction finish method should
 be employed whenever possible.
 
+## Session Management for Interaction Finish Methods {#security-sessions}
+
+When using an interaction finish method such as `redirect` or `push`, the client instance receives
+an unsolicited HTTP request from an unknown party, through the end-user's browser. The client
+instance needs to be able to successfully associate this incoming request with a specific pending
+grant request being managed by the client. If the client instance is not careful and precise about
+this, an attacker could associate their own session at the client instance with a stolen interaction
+response. The means of preventing this varies by the type of client and interaction methods in use.
+Some common patterns are enumerated here.
+
+If the end user interacts with the client instance through a web browser and the `redirect`
+interaction finish method is used, the client instance can ensure that the incoming HTTP request
+from the finish method is presented in the same browser session that the grant request was
+started in. This technique is particularly useful when the `redrect` interaction start mode
+is used as well. The client instance can then store the relevant pending grant information in the
+session, either in the browser storage directly (such as with a single-page application) or
+in an associated session store on a back-end server. In both cases, when the incoming request
+reaches the client instance, the session information can be used to ensure it's the same party
+that started the request is present as the request finishes.
```

Comment:
It depends on how the interactions are intended to happen. What you describe is the most common for sure, and I think it should be called out, but I didn't want this advice to make people assume you could only use redirect/redirect together. In fact, this is the best place to discuss what the different combinations would need for security considerations -- which is the impetus behind having this PR close out #364

Reply:
@pq2 Text in question has been updated and expanded, please take another look!

Reply:
I think the cases where a client instance is not able to strongly associate an incoming HTTP request with an established session with the end user are now well explained. However, I still miss a mention of the attack described in #364. I think we should either mention that this attack is possible if this association is not verified and the end user does not manage to determine from the information provided to him by the AS during the interaction that he does not want to authorize this client instance (which is a highly realistic assumption). Or we directly mention that this form of phishing attack can be prevented by ensuring that the same end user is present after the interaction.

Reply:
I tried to cover this in the later paragraphs of this section, and talking about the single-user case. Would you please propose text that you think would cover the attack more explicitly?

Reply:
How about something like this:
```suggestion
that started the request is present as the request finishes.

Ensuring that the same party that started a request is present when that request finishes can
prevent phishing attacks, where an attacker starts a request at an honest client instance and
tricks an honest RO into authorizing it. For example, if an honest end user (that also acts as the
RO) wants to start a request through a client instance controlled by the attacker, the attacker can
start a request at an honest client instance and then redirect the honest end user to the
interaction URI from the attackers session with the honest client instance. If the honest end user
then fails to realize that it is not authorizing the attacker-controlled client instance (with which
it started its request) but the honest client instance when interacting with the AS, the attacker's
session with the honest client instance would be authorized. This would give the attacker access to
the honest end user's resources that the honest client instance is authorized to access. However,
if after the interaction the AS redirects the honest end user back to the client instance whose
grant request the end user just authorized, the honest end user is redirected to the honest client
instance. The honest client instance can then detect that it is not the party that started the
request that is present, since the request at the honest client instance was started by the
attacker, which can prevent the attack. This is related to {{security-impersonation}}, because again
the attack can be prevented by the AS informing the user as much as possible about the client
instance that is to be authorized.
```

Reply:
This is good, thank you! I've included your text.